File: RecoverBinarySearchTree 99.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jun  5 19:03:12 2019

@author: haimingwd
"""

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    
    def inOrder(self, p):
        if not p:
            return
        self.inOrder(p.left)
        #print(p.val, self.prev.val)
        if not self.first and self.prev.val > p.val:
            self.first = self.prev
        if self.first and self.prev.val > p.val:
            self.second = p
        self.prev = p
        self.inOrder(p.right)
            
    def recoverTree(self, root: TreeNode) -> None:
        
        self.first, self.second, self.prev = None, None, TreeNode(float('-inf'))
        self.inOrder(root)
        self.first.val, self.second.val = self.second.val, self.first.val
        
        

File: Cpp/RecoverBinarySearchTree 99.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 
class Solution {
public:
    vector<TreeNode*> vec;
    
    void preorder(TreeNode* root){
        if(root){
            preorder(root->left);
            vec.push_back(root);
            preorder(root->right);
        }        
    }
    
    void recoverTree(TreeNode* root) {
        preorder(root);
        int n = vec.size();
        int i, j, k = 0;
        for(k = 0; k < n-1; k++){
            if(vec[k]->val > vec[k+1]->val){
                i = k;
                break;
            }
        }
        for(k = n-1; k > 0; k--){
            if(vec[k]->val < vec[k-1]->val){
                j = k;
                break;
            }
        }
        int temp = vec[i]->val;
        vec[i]->val = vec[j]->val;
        vec[j]->val = temp;
    }
};
